stock_available_to_promise_release/models/stock_move.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2019 Camptocamp (https://www.camptocamp.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+from odoo.addons import decimal_precision as dp
+from odoo.osv import expression
+from odoo.tools import float_compare
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    date_priority = fields.Datetime(
+        string="Priority Date",
+        index=True,
+        default=fields.Datetime.now,
+        help="Date/time used to sort moves to deliver first. "
+        "Used to calculate the ordered available to promise.",
+    )
+    ordered_available_to_promise = fields.Float(
+        "Ordered Available to Promise",
+        compute="_compute_ordered_available_to_promise",
+        digits=dp.get_precision("Product Unit of Measure"),
+        help="Available to Promise quantity minus quantities promised "
+        " to older promised operations.",
+    )
+    need_release = fields.Boolean()
+
+    @api.depends()
+    def _compute_ordered_available_to_promise(self):
+        for move in self:
+            move.ordered_available_to_promise = (
+                move._ordered_available_to_promise()
+            )
+
+    def _should_compute_ordered_available_to_promise(self):
+        return (
+            self.picking_code == "outgoing"
+            and not self.product_id.type == "consu"
+            and not self.location_id.should_bypass_reservation()
+        )
+
+    def _ordered_available_to_promise(self):
+        if not self._should_compute_ordered_available_to_promise():
+            return 0.
+        available = self.product_id.with_context(
+            location=self.warehouse_id.lot_stock_id.id
+        ).virtual_available
+        return max(
+            min(available - self._previous_promised_qty(), self.product_qty),
+            0.,
+        )
+
+    def _previous_promised_quantity_domain(self):
+        domain = [
+            ("need_release", "=", True),
+            ("product_id", "=", self.product_id.id),
+            ("date_priority", "<=", self.date_priority),
+            ("warehouse_id", "=", self.warehouse_id.id),
+        ]
```

Comment:
Ideally we should exclude all the moves planned in the future that you can resupply on time. For example, there is no reason a delivery planned 3 months is blocking quantity while you can resupply in 7 days.
Simple solution would be configuring an horizon for considering blocking moves. A more complex and accurate solution would be by looking at the warehouse product orderpoint planned date.
To support sales order commitment dates, something must be done

Reply:
We go for the horizon solution here